File: clang/test/SemaObjC/attr-suppress.m
// RUN: %clang_cc1 -fsyntax-only -fblocks %s -verify

#define SUPPRESS1 __attribute__((suppress))
#define SUPPRESS2(...) __attribute__((suppress(__VA_ARGS__)))

SUPPRESS1 int global = 42;

SUPPRESS1 void foo() {
  // expected-error@-1 {{'suppress' attribute only applies to variables and statements}}
  SUPPRESS1 int *p;

  SUPPRESS1 int a = 0; // no-warning
  SUPPRESS2()
  int b = 1; // no-warning
  SUPPRESS2("a")
  int c = a + b;                     // no-warning
  SUPPRESS2("a", "b") { b = c - a; } // no-warning

  SUPPRESS2("a", "b")
  if (b == 10)
    a += 4;              // no-warning
  SUPPRESS1 while (1) {} // no-warning
  SUPPRESS1 switch (a) { // no-warning
  default:
    c -= 10;
  }

  // GNU-style attributes and C++11 attributes apply to different things when
  // written like this.  GNU  attribute gets attached to the declaration, while
  // C++11 attribute ends up on the type.
  int SUPPRESS2("r") z;
  SUPPRESS2(foo)
  float f;
  // expected-error@-2 {{expected string literal as argument of 'suppress' attribute}}
}

union SUPPRESS2("type.1") U {
  // expected-error@-1 {{'suppress' attribute only applies to variables and statements}}
  int i;
  float f;
};

SUPPRESS1 @interface Test {
  // expected-error@-1 {{'suppress' attribute only applies to variables and statements}}
}
@property SUPPRESS2("prop") int *prop;
// expected-error@-1 {{'suppress' attribute only applies to variables and statements}}
- (void)bar:(int)x SUPPRESS1;
// expected-error@-1 {{'suppress' attribute only applies to variables and statements}}
@end

File: clang/test/SemaCXX/attr-suppress.cpp
// RUN: %clang_cc1 -std=c++11 -fsyntax-only %s -verify

[[gsl::suppress("globally")]];

namespace N {
[[gsl::suppress("in-a-namespace")]];
}

[[gsl::suppress("readability-identifier-naming")]] void f_() {
  int *p;
  [[gsl::suppress("type", "bounds")]] {
    p = reinterpret_cast<int *>(7);
  }

  [[gsl::suppress]] int x;       // expected-error {{'suppress' attribute takes at least 1 argument}}
  [[gsl::suppress()]] int y;     // expected-error {{'suppress' attribute takes at least 1 argument}}
  int [[gsl::suppress("r")]] z;  // expected-error {{'suppress' attribute cannot be applied to types}}
  [[gsl::suppress(f_)]] float f; // expected-error {{expected string literal as argument of 'suppress' attribute}}
}

union [[gsl::suppress("type.1")]] U {
  int i;
  float f;
};

// This doesn't really suppress anything but why not?
[[clang::suppress]];

namespace N {
[[clang::suppress("in-a-namespace")]];
} // namespace N

[[clang::suppress]] int global = 42;

[[clang::suppress]] void foo() {
  [[clang::suppress]] int *p;

  [[clang::suppress]] int a = 0;           // no-warning
  [[clang::suppress()]] int b = 1;         // no-warning
  [[clang::suppress("a")]] int c = a + b;  // no-warning
  [[clang::suppress("a", "b")]] b = c - a; // no-warning

  [[clang::suppress("a", "b")]] if (b == 10) a += 4; // no-warning
  [[clang::suppress]] while (true) {}                // no-warning
  [[clang::suppress]] switch (a) {                   // no-warning
  default:
    c -= 10;
  }

  int [[clang::suppress("r")]] z;
  // expected-error@-1 {{'suppress' attribute cannot be applied to types}}
  [[clang::suppress(foo)]] float f;
  // expected-error@-1 {{expected string literal as argument of 'suppress' attribute}}
}

class [[clang::suppress("type.1")]] V {
  int i;
  float f;
};

// FIXME: There's no good reason why we shouldn't support this case.
// But it doesn't look like clang generally supports such attributes yet.
class W : [[clang::suppress]] public V { // expected-error{{'suppress' attribute cannot be applied to a base specifier}}
};
